src/libxrpl/protocol/STAmount.cpp

```diff
@@ -1226,16 +1206,36 @@ multiply(STAmount const& v1, STAmount const& v2, Issue const& issue)
 
         return STAmount(v1.getFName(), minV * maxV);
     }
+    if (v1.holds<MPTIssue>() && v2.holds<MPTIssue>() && asset.holds<MPTIssue>())
+    {
+        std::uint64_t const minV = getMPTValue(v1) < getMPTValue(v2)
+            ? getMPTValue(v1)
+            : getMPTValue(v2);
+        std::uint64_t const maxV = getMPTValue(v1) < getMPTValue(v2)
+            ? getMPTValue(v2)
+            : getMPTValue(v1);
+
+        if (minV > 3000000000ull)  // sqrt(cMaxNative)
+            Throw<std::runtime_error>("Asset value overflow");
+
+        if (((maxV >> 32) * minV) > 2095475792ull)  // cMaxNative / 2^32
+            Throw<std::runtime_error>("Asset value overflow");
```

Comment:
Should these be comparing against `sqrt(maxMPTokenAmount)` and `maxMPTokenAmount / 2^32`, respectively? (I'll leave it to you to do the math.)